### DISCUSSION TITLE
menu items are greyed out when accessing the `menu item` from applescript

### DISCUSSION THREAD

**Discussion by ipatch** [rxhanson/Rectangle#346] [Ideas]
- [x] This is a new bug that hasn't been documented in a closed issue or in the Readme.

macOS version: latest version of mojave _10.14_

Rectangle version: Version 0.41 (46) _installed via homebrew_

Logs if applicable (In Rectangle menu, hold option, "View Logging..."): _nothing appears when viewing the logs_

coming from spectacle.app as i just found out it is no longer being actively developed removed spectacle.app and installed this app, ie. rectangle.app

i'd like to hold a key sequence down ie. _larger_ and _smaller_ as i move my windows around the screen, sometimes my browser window is small, and i'd like to make it larger, but i have to press the keyboard shortcut i have setup, ie. 

<kbd>⌘</kbd>+<kbd>⌥</kbd>+<kbd>➕</kbd> 

to make the window larger, and continually have to repeat the keyboard shortcut ie. i **can not** hold down such keyboard shortcut to make the window larger or smaller.

i thought i'd be clever and attempt to access the menu items from an applescript but most if not all menu items are greyed out when accessing them via an applescript, possibly due to the fact that i don't write applescripts to often, but this what the menu item looks like when running my applescript

<img width="1920" alt="tech my gh issue rectangle wm macos" src="https://user-images.githubusercontent.com/613805/105254702-c9b7be80-5b47-11eb-8421-fc37be3f6bd1.png">

i saved my applescript as an app bundle ie. a `.app` and setup a _karabiner-elements_ modification test, ie. holding a key down _right alt key_ which runs my applescript app bundle, but as mentioned in the image most if not all menu items are not accessible.

<details>
<summary> my applescript </summary>

```applescript
activate application "Rectangle"
tell application "System Events"
	tell process "Rectangle"
		tell (menu bar item 1 of menu bar 2)
			click -- need this line to access the menu bar items
			click menu item "Larger" of menu 1
		end tell
	end tell
end tell
```

</details>

**Comment by rxhanson** [comment]
By activating Rectangle in your AppleScript, you're bringing focus to Rectangle rather than the window you are trying to manipulate. There's no ui element / window for Rectangle to grab in this scenario, hence the greyed out items.

There are two potential feature requests here, rather than a bug.

1. Allow held keys to repeatedly execute larger/smaller actions.
2. Allow Rectangle to be programmatically managed via applescript

Request 1 isn't as trivial as it sounds, since Rectangle uses macOS keybindings (via MASShortcut), and the application is not notified of the down and up of keystrokes.

Request 2 has an open enhancement request. It's also a decent amount of work but is something I would be more open to adding in (PR welcome!)

Moving this to a discussion since I don't deem this as a bug

**Comment by dmland** [comment]
I know this request is a bit old, Keyboard Maestro lets me (somewhat slowly and hackily, but effectively) automate Rectangle, which isn't yet scriptable.

I use the Sixths menu, but navigating submenus is _meh_. Also, I didn't want to dedicate six keyboard shortcuts to a function I don't use that often, so I created a Keyboard Maestro macro that activates the Rectangle menu (by finding its icon in the menu bar and clicking on it), then sends a sequence of arrow keys to navigate the menu. For example, after opening the menu, it sends ↑(x6 to get to "Sixths ►"), then → and ⏎, which selects Top Left Sixth. Similarly ↑(x6), →, ↓(x4), ⏎ selects Bottom Center Sixth. It sounds a lot more complicated than it is, honestly.

All of the macros share the same keyboard shortcut. To choose the right one, Keyboard Maestro pops up a little window where you can type one or two characters to select it. By giving them reasonable names like "Top Left Sixth," I can type the shared keyboard shortcut, followed by "TR" for "**T**op **R**ight" or "BL" for "**B**ottom **L**eft" and so forth.

It's a bit slow because I had to add a 0.25s delay between keystrokes for it to be reliable and repeatable.

It works great. Keyboard Maestro's method of disambiguating macros with the same shortcut makes it work.
